### PR TITLE
Add shape metadata to DiskDataset

### PR DESCRIPTION
**Pull request by rbharath**
We've had issues with printing large datasets https://github.com/deepchem/deepchem/issues/1977 since `DiskDataset` doesn't have any way to check the shape of its underlying data without actually iterating through the underlying data on disk. This is really inefficient for large datasets and is also unintuitive (programs will freeze up when you're printing a large dataset).

In this PR, I introduce a strategy to deal with this by adding shape metadata to `DiskDataset`. This stores metadata about shapes in `self.metadata_df` and means that getting shapes for a `DiskDataset` will be very fast and can be done purely in memory. The downside to this is that it induces a change in the storage format for `DiskDataset`. I believe I've made sure that the legacy metadata format can still be constructed and worked with and added some tests for this. I've changed the `__str__`, `__repr__` and `get_shape` functions to rely on this metadata instead of looking at disk.

I think this PR is ready for a first round of review. We'll need to be a little careful on this one since we're changing the tried-and-true metadata for `DiskDataset`. I've added additional documentation to `DiskDataset` making notes about the metadata format and some usage example code to improve maintainability.

CC @peastman @nd-02110114 @ncfrey @seyonechithrananda @vsomnath

### CODE REVIEW

**Comment by rbharath** [comment]
There's a bit of cruft in here since I started from #2061's code. I'll rebase that out once #2061 is merged in

**Comment by peastman** [comment]
Why is the `legacy_metadata` argument needed?  When loading a dataset, we don't need the user to tell us.  We can figure that out just by checking the number of columns in the metadata.  And when creating a new dataset, the only reason to use the old format would be to create a dataset that could be read by older versions of DeepChem, but we already broke forward compatibility by switching from .joblib to .npy files.

File: deepchem/data/datasets.py [hunk 1]
```diff
@@ -1022,60 +1126,138 @@ def load_metadata(self):
...
       del metadata_df['basename']
-      save_metadata(tasks, metadata_df, self.data_dir)
+      DiskDataset._save_metadata(tasks, metadata_df, self.data_dir)
       return tasks, metadata_df
     raise ValueError("No Metadata Found On Disk")
 
   @staticmethod
-  def _construct_metadata(metadata_entries: List) -> pd.DataFrame:
-    """Construct a dataframe containing metadata.
+  def _save_metadata(tasks, metadata_df, data_dir):
```
Comment (peastman) [inline deepchem/data/datasets.py:1134]:
```python
   def _save_metadata(tasks: List[str], metadata_df: pd.DataFrame, data_dir: str) -> None:
```

File: deepchem/data/datasets.py [hunk 2]
```diff
@@ -1092,7 +1274,13 @@ def get_task_names(self) -> np.ndarray:
     return self.tasks
 
   def reshard(self, shard_size: int) -> None:
-    """Reshards data to have specified shard size."""
+    """Reshards data to have specified shard size.
+
+    Note
+    ----
+    If this `DiskDataset` is in `legacy_metadata` format, reshard will
+    maintain legacy metadata format.
```
Comment (peastman) [inline deepchem/data/datasets.py:1282]: Is there any reason not to upgrade it to the new format?  That will give better performance.

File: deepchem/data/datasets.py [hunk 3]
```diff
@@ -1473,8 +1665,33 @@ def from_numpy(X: np.ndarray,
...
+      labels array
+    w: Optional[np.ndarray], optional (default None)
+      weights array
+    ids: Optional[np.ndarray], optional (default None)
+      identifiers array
+    tasks: Optional[Sequence], optional (default None)
+      Tasks in this dataset
+    data_dir: Optional[str], optional (default None)
+      The directory to write this dataset to. If none is specified, will use
+      a temporary dataset instead.
```
Comment (peastman) [inline deepchem/data/datasets.py:1686]: I think you meant "a temporary directory instead"?

File: deepchem/data/datasets.py [hunk 4]
```diff
@@ -957,24 +961,115 @@ def merge(datasets: Sequence[Dataset]) -> "NumpyDataset":
...
+    Whether this `DiskDataset` uses legacy format.
+
+  Note
+  ----
+  `DiskDataset` originally had a simpler metadata format without shape
+  information. Older `DiskDataset` objects had metadata files with columns
+  `('ids', 'X', 'y', 'w') and not additional shape columns. `DiskDataset`
+  maintains backwards compatibility with this older metadata format, but we
+  recommend for performance reasons not using legacy metadata for new
+  projects.
```
Comment (peastman) [inline deepchem/data/datasets.py:1037]: The improved documentation looks great!

File: deepchem/data/datasets.py (hunk 3 above)
Comment (rbharath) [reply deepchem/data/datasets.py:1686]: Good catch! Will fix

File: deepchem/data/datasets.py (hunk 1 above)
Comment (rbharath) [reply deepchem/data/datasets.py:1134]: Good catch! Will fix

File: deepchem/data/datasets.py (hunk 2 above)
Comment (rbharath) [reply deepchem/data/datasets.py:1282]: Good point. I'll upgrade it to the new format

File: deepchem/data/datasets.py (hunk 4 above)
Comment (rbharath) [reply deepchem/data/datasets.py:1037]: Thanks!

**Comment by rbharath** [comment]
Good point that we can infer whether it's legacy or not! I've gone ahead and removed it from the constructor.

I think it's useful to keep around the option to create datasets in legacy format to facilitate testing for backwards compatibility with reading legacy datasets, so I've left the option in `create_dataset` and `from_numpy`. I've addressed all open comments so I think this is ready for another round of review :)

**Comment by peastman** [comment]
I agree with @ncfrey on this, but I'd go even further.  The new format is clearer superior.  There's no reason for a user to want a dataset with the old format.  So why provide an option that users should never actually use?

If we're just concerned about testing, let's stick a small dataset using the old format in the `tests` directory and have a test make sure it still loads correctly.

File: deepchem/data/datasets.py [hunk 5]
```diff
@@ -957,24 +961,138 @@ def merge(datasets: Sequence[Dataset]) -> "NumpyDataset":
...
+  `DiskDataset._construct_metadata` and saved to disk by
+  `DiskDataset._save_metadata`. The metadata itself consists of a csv file
+  which has columns `('ids', 'X', 'y', 'w', 'ids_shape', 'X_shape', 'y_shape',
+  'w_shape')`. `tasks.json` consists of a list of task names for this dataset.
+
+  The actual data is stored in `.npy` files (numpy array files) of the form
+  'shard-0-X.npy', 'shard-0-y.npy', etc.
+
+  The basic structure of `DiskDataset` is quite robust and will likely serve
+  you will for datasets up to about 100 GB or larger. However note that
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/data/datasets.py:1002]: Good catch! Will fix

File: deepchem/data/datasets.py [hunk 6]
```diff
@@ -1026,30 +1150,66 @@ def load_metadata(self):
...
-      save_metadata(tasks, metadata_df, self.data_dir)
+      DiskDataset._save_metadata(tasks, metadata_df, self.data_dir)
       return tasks, metadata_df
     raise ValueError("No Metadata Found On Disk")
 
   @staticmethod
-  def _construct_metadata(metadata_entries: List) -> pd.DataFrame:
+  def _save_metadata(tasks: Optional[Sequence], metadata_df: pd.DataFrame,
+                     data_dir: str) -> None:
+    """Saves the metadata for a DiskDataset
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/data/datasets.py:1160]: Good suggestion! Will do

File: deepchem/data/datasets.py [hunk 7]
```diff
@@ -1026,30 +1150,66 @@ def load_metadata(self):
...
+      tasks = tasks.tolist()
+    metadata_filename = os.path.join(data_dir, "metadata.csv.gzip")
+    tasks_filename = os.path.join(data_dir, "tasks.json")
+    with open(tasks_filename, 'w') as fout:
+      json.dump(tasks, fout)
+    metadata_df.to_csv(metadata_filename, index=False, compression='gzip')
+
+  @staticmethod
+  def _construct_metadata(metadata_entries: List,
+                          legacy_metadata: bool = False) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/data/datasets.py:1183]: Good catch, will fix!

File: deepchem/data/datasets.py [hunk 8]
```diff
@@ -957,24 +961,138 @@ def merge(datasets: Sequence[Dataset]) -> "NumpyDataset":
...
     """
     self.data_dir = data_dir
 
     logger.info("Loading dataset from disk.")
     self.tasks, self.metadata_df = self.load_metadata()
+    if len(self.metadata_df.columns) == 4 and list(
+        self.metadata_df.columns) == ['ids', 'X', 'y', 'w']:
+      logger.info(
+          "Detected legacy metatadata on disk. You can upgrade from legacy metadata to the more efficient current metadata by resharding this dataset."
+      )
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/data/datasets.py:1076]: Good suggestion, will add!

**Comment by rbharath** [comment]
@peastman @ncfrey Good points! I've removed support for writing legacy metadata and have added two small test datasets written in the legacy format to the test suite. 

To facilitate testing, I also added a new function `DiskDataset.copy()` which makes a copy of a `DiskDataset` and fixed a bug in `DiskDataset.move`. I've also added tests for moving and copying `DiskDataset`.

**Comment by rbharath** [comment]
I've addressed all open comments and only failing tests are the known ones with pytorch-dataset so going to go ahead and merge this in

**Pull request merged**